Handle missing OS minor version gracefully

MichalHe opened the pull request:
Handle missing OS minor version gracefully, instead of crashing at various points of execution: when checking the version format in the CLI, or when using config.version library functions.

The proposed solution adds a special object into  `upgrade_paths.json` called `_virtual_versions` (under the upgrade paths for `CentOS`). When running on CentOS, we do not require the `MAJOR.MINOR` version format anymore. Furthermore, whenever the (old) code uses `config.version` library functions, we use minor versions defined in `_virtual_versions` to use a complete version for version comparisons. This approach guarantees predictable and correct behavior of version-specific checks. Other solutions were discussed on a team cabal, and we found edge cases when the code could exhibit unexpected behavior, such as version-specific workaround code not being executed (or being executed) at all, although it should be. Therefore, this solution, although a bit labor intensive, seems like the only reliable way to make the already written code work on CentOS.

**TODO**:
- [x] Check that version.matches_version and other affected library functions behave as expected, i.e., they use virtual versions

Jira-ref: RHEL-80334

Commit:
feat(command_utils): allow missing minor/major version for CentOS

CentOS versioning consists only of the major version number. Therefore,
in case we are running on CentOS, we do not want to enforce version to
have the format MAJOR.MINOR. Rather, we want to check that the version
is of the form MAJOR. This patch introduces the necessary
infrastructure, allowing easy extensibility of version format checking
for also other distributions. In case we do not know what version
format should a distro use, we default to MINOR.MAJOR.

Jira-ref: RHEL-80334

Commit:
refactor(versions/config): rename _simple_versions

Rename the function _simple_versions to _are_comparison_operators_used
with **negated semantics**. The new name should make it clear what is
being checked in the version list.

Commit:
fix(lib/version): broken _validate_version on RHEL 10

The validate version used to split a given version on `.` and then
check that every fragment is a digit, which is false for RHEL 10
since `10` is not a digit.

pirat89 commented on repos/system_upgrade/common/libraries/config/version.py:
```diff
@@ -141,17 +141,17 @@ def _version_to_tuple(version):
 
 
-def _simple_versions(versions):
+def _are_comparision_operators_used(versions):
```
do you mean `comparison` ? :)

MichalHe replied:
Should, be fixed now

pirat89 commented on commands/command_utils.py:
```diff
@@ -16,26 +18,52 @@
+    TARGET = 'target'
+
+
+class RHELDistros(str, Enum):
```
I would not use the `RHELDistros` name in this context. Let's use rather something like `DistroIDs`

MichalHe replied:
Fixed, thanks.

pirat89 commented on commands/command_utils.py:
```diff
@@ -71,15 +99,33 @@ def get_upgrade_flavour():
 
 
-def _retrieve_os_release_contents(_os_release_path='/etc/os-release'):
+@functools.cache
```
@MichalHe `functools.cache` is defined since Python 3.9, but the code is executed with Python 3.6 as well:
* https://docs.python.org/3/library/functools.html#functools.cache

Is there a high enough number of calls to have a benefit from using the cache wrapper at all?

MichalHe replied:
Thank you for noticing, I have dropped the use of `functools`.

> Is there a high enough number of calls to have a benefit from using the cache wrapper at all?

Since it prevents us from accessing the filesystem (systemcalls are usually expensive), then the cost of the wrapper is small. I somehow think that the performance of the tool (how fast can I iterate) is something that I don' enjoy about the development experience, so I try to take into consideration also that aspect of the code.

Commit:
upgrade_paths: add information about centos virtual versions

Add additional field to upgrade paths for CentOS that maps CentOS
versions consisting only of a major version number to corresponding
RHEL versions of the form MAJOR.MINOR ("virtual version"). This
information is inteded to be used by leapp to in version checks
to obtain a MAJOR.MINOR version so that code written for RHEL can work
as expected also on CentOS.

Jira-ref: RHEL-80334

Commit:
models(ipuconfig): extend Version class to contain virtual versions

Add virtual_{source,target}_version fields to the Version model. These
fields store a virtual MAJOR.MINOR CentOS version so that
version-specific checks originally written for RHEL can work as expected
also on CentOS. On non-CentOS system, the value of these fields should
be the same as source/target versions. The ipuworkflowconfig actor is
modified accordingly to populate the newly added fields.

Jira-ref: RHEL-80334

Commit:
libs(common, version): autocorrect centos versions into MAJOR.MINOR

When executing version.matches_version on CentOS, autocorrect versions
of the form MAJOR into MAJOR.MINOR using virtual versions available in
IPU configuration. Autocorrection is implemented only if a used version
matches source/target version. For example, if match_list contains '8'
on a CentOS 8 system, the version will be autocorrected to '8.10'.
However, if a version that does not match the source/target version is
present, it will be left untouched.

Jira-ref: RHEL-80334

MichalHe commented:
@oamg/developers review please

evgeni commented on commands/command_utils.py:
```diff
@@ -16,26 +18,52 @@
+    TARGET = 'target'
+
+
+class DistroIDs(str, Enum):
+    RHEL = 'rhel'
+    CENTOS = 'centos'
+
+
+_DISTRO_VERSION_FORMATS = {
+    DistroIDs.RHEL: VersionFormats.MAJOR_MINOR,
+    DistroIDs.CENTOS: VersionFormats.MAJOR_ONLY,
+}
+"""
+Maps distro ID to the expected OS version format.
+
+If a distro is not listed in the dictionary, then VersionFormats.MAJOR_MINOR
+is used as a default.
+"""
```
How can consumers extend these lists (without forking the codebase)?
While this probably won't matter much (for Alma and friends) if the fallback is MAJOR_MINOR, but still curious :)

MichalHe replied:
Hmm, community projects can always create an upstream PR extending the list, and there should be no problem in merging such a change. Not sure whether implementing something that provides extensibility is worth the effort, since it is not as simple as 'writing an actor' due to the code being executed very early.

pirat89 commented on commands/command_utils.py:
```diff
@@ -71,15 +99,32 @@ def get_upgrade_flavour():
+            if '=' not in line:
+                continue
+
+            key_val_pair = line.strip().split('=', 1)
+
+            if strip_double_quotes:
+                key_val_pair = strip_quotes_from_field_value(key_val_pair)
+
+            os_release_contents[key_val_pair[0]] = key_val_pair[1]
```
nitpick, but I would do rather something like this:
```suggestion
            key, val = line.strip().split('=', 1)

            if strip_double_quotes:
                val = strip_quotes_from_field_value(val)

            os_release_contents[key] = val
```

The nested `strip_quotes_from_field_value` function would need to be updated as well then. Possibly it could be also renamed to `strip_quotes`.

pirat89 commented on repos/system_upgrade/common/files/upgrade_paths.json:
```diff
@@ -21,6 +21,11 @@
+        },
+        "_virtual_versions": {
+            "8": "8.10",
+            "9": "9.5",
+            "10": "10.0"
```
Can be changed later, but based on the merge time, we can virtual versions already 9.6, 10.0. It would be possible even 9.7, 10.1 as Centos Stream contains already changes for new minor releases, we can stay with 9.6, 10.0.

This does not block the merge, it's a nitpick now.

MichalHe replied:
I have modified virtual version for CentOS 9 to 9.6. Virtual version for CentOS 10 is unchanged.

pirat89 commented on repos/system_upgrade/common/actors/ipuworkflowconfig/libraries/ipuworkflowconfig.py:
```diff
@@ -92,8 +93,7 @@ def load_upgrade_paths_definitions(paths_definition_file):
 
-def load_raw_upgrade_paths_for_distro_and_flavour(distro_id, flavour, paths_definition_file='upgrade_paths.json'):
-    all_definitions = load_upgrade_paths_definitions(paths_definition_file)
+def extract_paths_upgrade_paths_for_distro_and_flavour(all_definitions, distro_id, flavour):
```
nitpick, but why `extract_paths_upgrade_paths_for_distro_and_flavour` instead of `extract_upgrade_paths_for_distro_and_flavour`? Reading the name, it's a little bit confusing

MichalHe replied:
There is no reason, I'm just blind :). Should be fixed.

pirat89 commented on repos/system_upgrade/common/actors/ipuworkflowconfig/libraries/ipuworkflowconfig.py:
```diff
@@ -117,6 +117,39 @@ def construct_models_for_paths_matching_source_major(raw_paths, src_major_versio
+
+    centos_upgrade_paths = all_upgrade_path_defs.get('centos', {})
+    if not centos_upgrade_paths:
+        raise StopActorExecutionError('There are no upgrade paths are defined for CentOS.')
```
typo:

```suggestion
        raise StopActorExecutionError('There are no upgrade paths defined for CentOS.')
```

MichalHe replied:
Thank you, should be fixed.

pirat89 commented on repos/system_upgrade/common/models/ipuconfig.py:
```diff
@@ -33,6 +33,21 @@ class Version(Model):
+
+    CentOS has version of the form MAJOR, lacking the minor version number. The
+    `virtual_source_version` value is obtained by combining CentOS major
+    version number with a minor version number stored internally in leapp.
```
This could be confusing as there is not such a value stored in leapp (project).

```suggestion
    version number with a minor version number stored internally in the upgrade_paths.json file.
```

pirat89 reviewed (approved):
thanks! lgtm. I will do yet some manual testing before merging but everything seems right and it's covered by unit tests. The failed tests do are not releated to this change.

MichalHe commented:
@pirat89 Ups, I wanted to merge PR1378, but I have not noticed that I switched to the wrong tab (with this PR). Sorry, I did not intent to do so. Nothing should be seriously broken (I hope). Let me know if you discover any issues, I will promptly do a follow up PR.

pirat89 commented:
@MichalHe  np :D I just wanted to merge it now.
